addons/sale_stock/models/sale_order.py

```diff
@@ -35,31 +35,27 @@ def _default_warehouse_id(self):
     picking_ids = fields.One2many('stock.picking', 'sale_id', string='Transfers')
     delivery_count = fields.Integer(string='Delivery Orders', compute='_compute_picking_ids')
     procurement_group_id = fields.Many2one('procurement.group', 'Procurement Group', copy=False)
-    effective_date = fields.Date("Effective Date", compute='_compute_effective_date', store=True, help="Completion date of the first delivery order.")
-    expected_date = fields.Datetime( help="Delivery date you can promise to the customer, computed from the minimum lead time of "
-                                          "the order lines in case of Service products. In case of shipping, the shipping policy of "
+    expected_date = fields.Datetime(help="Delivery date you can promise to the customer, computed from the minimum lead time of "
+                                          "the order lines in case of Service products.\n In case of shipping, the shipping policy of "
                                           "the order will be taken into account to either use the minimum or maximum lead time of "
-                                          "the order lines.")
+                                          "the order lines.",
+                                    default=fields.Datetime.now)
+    is_delivered = fields.Boolean(compute='_compute_is_delivered', help="Technical field used to indicates if all of the products in the SO were fully delivered")
 
-    @api.depends('picking_ids.date_done')
-    def _compute_effective_date(self):
-        for order in self:
-            pickings = order.picking_ids.filtered(lambda x: x.state == 'done' and x.location_dest_id.usage == 'customer')
-            dates_list = [date for date in pickings.mapped('date_done') if date]
-            order.effective_date = min(dates_list).date() if dates_list else False
+    def _update_expected_date(self):
+        self.ensure_one()
+        if self.picking_policy == 'one':
+            super(SaleOrder, self)._update_expected_date(extremum=max)
+        else:
+            super(SaleOrder, self)._update_expected_date(extremum=min)
 
-    @api.depends('picking_policy')
-    def _compute_expected_date(self):
-        super(SaleOrder, self)._compute_expected_date()
-        for order in self:
-            dates_list = []
-            confirm_date = fields.Datetime.from_string(order.date_order if order.state in ['sale', 'done'] else fields.Datetime.now())
-            for line in order.order_line.filtered(lambda x: x.state != 'cancel' and not x._is_delivery()):
-                dt = confirm_date + timedelta(days=line.customer_lead or 0.0)
-                dates_list.append(dt)
-            if dates_list:
-                expected_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
-                order.expected_date = fields.Datetime.to_string(expected_date)
+    @api.depends('picking_ids.state')
+    def _compute_is_delivered(self):
```

Comment:
I would rather check on the SO line qty_delivered. Because if the customer do a return, he will have picking done but he will not be able to update a new expected date for the product to resend. To check with OMA